content/posts/2021/reana-client-0.7.3.md

```diff
@@ -0,0 +1,346 @@
+
+Let us use the previous workflow example to depict this
+functionality. In that workflow, the environment used was
+`python:3.9-slim`, which is properly tagged and exists in
+Docker Hub. We are going to modify it and use a non-existing image
+instead, for example `foo:bar`.
+
+```diff
+   specification:
+     steps:
+       - name: run-script
+-        environment: 'python:3.9-slim'
++        environment: 'foo:bar'
+         commands:
+           - python "${script_path}" "${sleeptime}"
+```
+
+Let us call the validator, passing the environments option:
+
+```console
+$ reana-client validate -f reana.yaml --environments
+==> Verifying REANA specification file... my-analysis/reana.yaml
+  -> SUCCESS: Valid REANA specification file.
+==> Verifying workflow parameters and commands...
+==> Verifying environments in REANA specification file...
+  -> SUCCESS: Environment image foo:bar has the correct format.
+  -> WARNING: Environment image foo:bar does not exist locally.
+  -> WARNING: Environment image foo:bar does not exist in Docker Hub: "Resource not found"
+  -> ERROR: Environment image foo:bar does not exist locally or remotely.
+```
+
+Let us analyse the environment validation output. First, we get a
+success message informing us that the image has the correct format.
+This is because, as expected, it has the image name and the tag, so
+in that regard it is correct. Then we get two warnings telling us
+that the image was not found either locally nor in the Docker Hub
+registry.
+As a consequence of this, we can see the last error message, telling
+us that the image does not exist, thus making the validation fail and
+stop its execution.
+
+Let us change the environment to use a valid one. We are going to
+revert back to the `python` image but using the `latest` tag instead:
+
+```diff
+   specification:
+     steps:
+       - name: run-script
+-        environment: 'foo:bar'
++        environment: 'python:latest'
+         commands:
+           - python "${script_path}" "${sleeptime}"
+```
+
+This is something that we strongly discourage as the `latest` tag can
+evolve in time, i.e. the image that is pointing today might, and
+probably will be different in the future. Let us run the validator
+again:
+
+```console
+$ reana-client validate -f reana.yaml --environments
+==> Verifying REANA specification file... my-analysis/reana.yaml
+  -> SUCCESS: Valid REANA specification file.
+==> Verifying workflow parameters and commands...
+==> Verifying environments in REANA specification file...
+  -> WARNING: Using 'latest' tag is not recommended in python environment image.
+  -> WARNING: Environment image python:latest does not exist locally.
+  -> SUCCESS: Environment image python:latest exists in Docker Hub.
+  -> WARNING: UID/GIDs validation skipped, specify `--pull` to enable it.
+```
+
+As part of the environment verification, we get a message warning us
+about the usage of the `latest` tag. Additionally, we have two checks to
```

Comment:
Additionally, we see two checks...

(we are kind of "discovering the features" together with the reader, so the story is presented from that unfolding perspective, hence "see" rather than "have")